src/core/qgsfeaturefiltermodel.cpp

```diff
@@ -0,0 +1,306 @@
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsfeaturefiltermodel.h"
+#include "qgsfeaturefiltermodel_p.h"
+
+QgsFeatureFilterModel::QgsFeatureFilterModel( QObject *parent )
+  : QAbstractItemModel( parent )
+  , mSourceLayer( nullptr )
+{
+  mReloadTimer.setInterval( 100 );
+  mReloadTimer.setSingleShot( true );
+  connect( &mReloadTimer, &QTimer::timeout, this, &QgsFeatureFilterModel::scheduledReload );
+}
+
+QgsVectorLayer *QgsFeatureFilterModel::sourceLayer() const
+{
+  return mSourceLayer;
+}
+
+void QgsFeatureFilterModel::setSourceLayer( QgsVectorLayer *sourceLayer )
+{
+  if ( mSourceLayer == sourceLayer )
+    return;
+
+  mSourceLayer = sourceLayer;
+  reload();
+  emit sourceLayerChanged();
+}
+
+QString QgsFeatureFilterModel::displayExpression() const
+{
+  return mDisplayExpression;
+}
+
+void QgsFeatureFilterModel::setDisplayExpression( const QString &displayExpression )
+{
+  if ( mDisplayExpression == displayExpression )
+    return;
+
+  mDisplayExpression = displayExpression;
+  reload();
+  emit displayExpressionChanged();
+}
+
+QString QgsFeatureFilterModel::filterValue() const
+{
+  return mFilterValue;
+}
+
+void QgsFeatureFilterModel::setFilterValue( const QString &filterValue )
+{
+  if ( mFilterValue == filterValue )
+    return;
+
+  mFilterValue = filterValue;
+  reload();
+  emit filterValueChanged();
+}
+
+QString QgsFeatureFilterModel::filterExpression() const
+{
+  return mFilterExpression;
+}
+
+void QgsFeatureFilterModel::setFilterExpression( const QString &filterExpression )
+{
+  if ( mFilterExpression == filterExpression )
+    return;
+
+  mFilterExpression = filterExpression;
+  reload();
+  emit filterExpressionChanged();
+}
+
+bool QgsFeatureFilterModel::loading() const
+{
+  return mGatherer;
+}
+
+QModelIndex QgsFeatureFilterModel::index( int row, int column, const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent )
+  return createIndex( row, column, nullptr );
+}
+
+QModelIndex QgsFeatureFilterModel::parent( const QModelIndex &child ) const
+{
+  Q_UNUSED( child )
+  return QModelIndex();
+}
+
+int QgsFeatureFilterModel::rowCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent );
+
+  return mEntries.size();
+}
+
+int QgsFeatureFilterModel::columnCount( const QModelIndex &parent ) const
+{
+  Q_UNUSED( parent )
+  return 1;
+}
+
+QVariant QgsFeatureFilterModel::data( const QModelIndex &index, int role ) const
+{
+  if ( !index.isValid() )
+    return QVariant();
+
+  switch ( role )
+  {
+    case Qt::DisplayRole:
+    case ValueRole:
+      return mEntries.value( index.row() ).value;
+
+    case FeatureIdRole:
+      return mEntries.value( index.row() ).id;
+  }
+
+  return QVariant();
+}
+
+void QgsFeatureFilterModel::updateCompleter()
+{
+  QVector<Entry> entries = mGatherer->entries();
```

Comment:
Should any of this logic be ported to QgsFieldValuesLineEdit?

Reply:
I've never worked with that... You tell me?

Reply:
The situation is quite similar - gathering field values in a background thread. But it uses just a simple QStringList completer. That's why I'm wondering what this extra logic does and if it would improve the other widget too.

Reply:
It manages entries in the form `key: displayExpression` and not just simple values.